mapflow/functional/layer_utils.py

```diff
@@ -274,53 +275,38 @@ def add_layer(self, layer: Optional[QgsMapLayer]) -> None:
             self.layer_group.insertLayer(order, layer)
             self.layer_group.setExpanded(True)
         else:  # assume user opted to not use a group, add layers as usual
             self.project.addMapLayer(layer)
 
-    def add_basemap_layer(self, basemap, layer):
-        # Remove layer if already exists
-        for l in self.iface.mapCanvas().layers():
-            if basemap in l.dataProvider().dataSourceUri():
-                QgsProject.instance().removeMapLayers([l.id()])
-        # Append each basemap layer to a list and add new ones at the top of a group
-        basemaps = []
-        for l in self.iface.mapCanvas().layers():
-            if "{z}" or "http" in l.dataProvider().dataSourceUri():
-                basemaps.append(l.id())
-            other_preview_layer_id = basemaps[-1]
-            other_preview_layer = QgsProject.instance().layerTreeRoot().findLayer(other_preview_layer_id)
-            parent_group = other_preview_layer.parent()
-            index = parent_group.children().index(other_preview_layer)
-            self.add_layer_by_order(order = index, layer = layer)
-            return
-        self.add_layer_by_order(layer=layer, order=-1)
-        return
+    def add_preview_layer(self, preview_layer, preview_list):        
+        # If new preview uri is in a list of previously added previes
+        if preview_layer.dataProvider().dataSourceUri() in preview_list:
+            # Delete it from the list and from the layer tree   
+            preview_list.remove(preview_layer.dataProvider().dataSourceUri())   
+            for l in self.iface.mapCanvas().layers():
+                if preview_layer.dataProvider().dataSourceUri() == l.dataProvider().dataSourceUri():
+                    QgsProject.instance().removeMapLayers([l.id()])
+        # If preview list is empty
+        if len(preview_list) == 0:
+            # Add preview to the bottom and its uri to the list
+            self.add_layer(layer = preview_layer, order=-1)
+            preview_list.append(preview_layer.dataProvider().dataSourceUri())
+        else:
+            # Get the last value from previews list
+            upper_preview_url = preview_list[-1]
+            # Compare it to layer tree to find a specific layer id and its index
+            for l in self.iface.mapCanvas().layers():
+                if upper_preview_url == l.dataProvider().dataSourceUri():
+                    upper_preview_id = l.id()
+                    upper_preview = QgsProject.instance().layerTreeRoot().findLayer(upper_preview_id)
+                    parent_group = upper_preview.parent()
+                    index = parent_group.children().index(upper_preview)
+                    # Add preview to the found position and its uri to the list
+                    self.add_layer(layer = preview_layer, order = index)
+                    preview_list.append(preview_layer.dataProvider().dataSourceUri())
```

Comment:
I would rather keep the list of the layers, which will ensure that we are talking about the same layers, not just something with the same URL:

```preview_list.append(preview_layer)```

then I could have a function like
```
def retrieve_preview(url, preview_list):
    for layer in preview_list:
        if not layer in project:
             # remove it from preview_list
        if (layer.dataSourceUri == url):
            return layer
     return create_new_layer(url)

```
Of course, we will need to implement "layer in project" and "create_new_layer"

Reply:
I understand the concern with uri's. But the problem is that layer by itself is still not a unique object:
![image](https://github.com/Geoalert/mapflow-qgis/assets/136620754/e655b172-a9f9-480e-bd28-1e93a6893a60)
Plus, for some reason, when trying to operate layers in a list (and not layer's name, id or uri, which are strings), they append as an empty string:
![image](https://github.com/Geoalert/mapflow-qgis/assets/136620754/64e58ef4-68a4-4e92-91bb-3996b63af0f2)
(whether I add _str_ or not), which makes it imposible to use. 
We could change it to layer.name() thougth and it would still be almost the same level of non-uniqueness.
To be honest, I don't know, how to solve this issue..